src/gems/simulation/extra_output.py

```diff
@@ -0,0 +1,155 @@
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict
+
+from gems.expression import evaluate
+from gems.expression.evaluate import EvaluationError, ValueProvider
+from gems.expression.expression import ExpressionNode
+from gems.study.data import ComponentParameterIndex, TimeScenarioIndex
+
+
+@dataclass
+class ExtraOutput:
+    """Stores evaluated outputs (from ExpressionNodes), not solver variables."""
+
+    name: str
+    values: Dict[TimeScenarioIndex, float] = field(default_factory=dict)
+
+    def set(self, t: int, s: int, value: float) -> None:
+        self.values[TimeScenarioIndex(t, s)] = value
+
+    def get(self, t: int, s: int) -> float | None:
+        return self.values.get(TimeScenarioIndex(t, s))
+
+
+def evaluate_all_extra_outputs(
+    problem: Any, component_lookup: Callable[[str], Any]
+) -> Dict[str, Dict[str, ExtraOutput]]:
+    """
+    Evaluate all model-defined extra outputs for every component in the network.
+    """
+    if problem is None or getattr(problem, "context", None) is None:
+        return {}
+
+    database = getattr(problem.context, "database", None)
+    if database is None:
+        print("[WARN] No database found in problem context; extra outputs skipped.")
+        return {}
+
+    results: Dict[str, Dict[str, ExtraOutput]] = {}
+
+    for cmp in problem.context.network.all_components:
+        model_def = cmp.model
+        if not model_def or not model_def.extra_outputs:
+            continue
+
+        comp_obj = component_lookup(cmp.id)
+        extra_results = evaluate_extra_outputs_for_a_component(comp_obj, problem)
+
+        if extra_results:
+            results[cmp.id] = extra_results
+
+    return results
+
+
+def evaluate_extra_outputs_for_a_component(
+    component: Any, problem: Any
+) -> Dict[str, ExtraOutput]:
+    """
+    Evaluate one component's model-defined extra outputs using solver variables as context.
+    """
+    results: Dict[str, ExtraOutput] = {}
+    model_def = getattr(component, "model", None)
```

Comment:
With proper typing, is this necessary ? With a simple `if component is None: return results` you should be able to get rid of all `getattr` if favour of the more readable `component.<attribute>` calls